// ff-overlay/esm5/lib/ff-overlay.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ApplicationRef, ComponentFactoryResolver, Injectable, Injector, RendererFactory2 } from '@angular/core';
import { FFOverlayComponent } from './ff-overlay.component';
import * as i0 from "@angular/core";
var FFOverlayService = /** @class */ (function () {
    function FFOverlayService(componentFactoryResolver, appRef, injector, rendererFactory) {
        this.componentFactoryResolver = componentFactoryResolver;
        this.appRef = appRef;
        this.injector = injector;
        this.renderer = rendererFactory.createRenderer(null, null);
    }
    /**
     * @private
     * @param {?} component
     * @param {?=} target
     * @return {?}
     */
    FFOverlayService.prototype.appendComponent = /**
     * @private
     * @param {?} component
     * @param {?=} target
     * @return {?}
     */
    function (component, target) {
        var _this = this;
        /** @type {?} */
        var componentRef = this.componentFactoryResolver
            .resolveComponentFactory(component)
            .create(this.injector);
        this.appRef.attachView(componentRef.hostView);
        /** @type {?} */
        var domElem = (/** @type {?} */ (((/** @type {?} */ (componentRef.hostView)))
            .rootNodes[0]));
        if (target) {
            this.renderer.appendChild(target, domElem);
        }
        else {
            this.renderer.appendChild(document.body, domElem);
        }
        return {
            componentRef: componentRef, destroy: (/**
             * @return {?}
             */
            function () {
                _this.appRef.detachView(componentRef.hostView);
                componentRef.destroy();
            })
        };
    };
    /**
     * @return {?}
     */
    FFOverlayService.prototype.getOverlay = /**
     * @return {?}
     */
    function () {
        if (!this._overlay) {
            this.init();
        }
        return this._overlay.componentRef.hostView.rootNodes[0];
    };
    /**
     * @param {?} domElement
     * @return {?}
     */
    FFOverlayService.prototype.appendChild = /**
     * @param {?} domElement
     * @return {?}
     */
    function (domElement) {
        this.renderer.appendChild(this.getOverlay(), domElement);
    };
    /**
     * @param {?} domElement
     * @return {?}
     */
    FFOverlayService.prototype.removeChild = /**
     * @param {?} domElement
     * @return {?}
     */
    function (domElement) {
        this.renderer.removeChild(this.getOverlay(), domElement);
        if (!this.getOverlay().children.length) {
            this.destroy();
        }
    };
    /**
     * @param {?=} target
     * @return {?}
     */
    FFOverlayService.prototype.init = /**
     * @param {?=} target
     * @return {?}
     */
    function (target) {
        if (!this._overlay) {
            this._overlay = this.appendComponent(FFOverlayComponent, target);
        }
    };
    /**
     * @return {?}
     */
    FFOverlayService.prototype.destroy = /**
     * @return {?}
     */
    function () {
        this._overlay && this._overlay.destroy();
        this._overlay = null;
    };
    /**
     * @return {?}
     */
    FFOverlayService.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.destroy();
    };
    FFOverlayService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    FFOverlayService.ctorParameters = function () { return [
        { type: ComponentFactoryResolver },
        { type: ApplicationRef },
        { type: Injector },
        { type: RendererFactory2 }
    ]; };
    /** @nocollapse */ FFOverlayService.ngInjectableDef = i0.defineInjectable({ factory: function FFOverlayService_Factory() { return new FFOverlayService(i0.inject(i0.ComponentFactoryResolver), i0.inject(i0.ApplicationRef), i0.inject(i0.INJECTOR), i0.inject(i0.RendererFactory2)); }, token: FFOverlayService, providedIn: "root" });
    return FFOverlayService;
}());
export { FFOverlayService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FFOverlayService.prototype._overlay;
    /**
     * @type {?}
     * @private
     */
    FFOverlayService.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    FFOverlayService.prototype.componentFactoryResolver;
    /**
     * @type {?}
     * @private
     */
    FFOverlayService.prototype.appRef;
    /**
     * @type {?}
     * @private
     */
    FFOverlayService.prototype.injector;
}
