// ff-resize-sensor/esm5/lib/ff-resize-sensor.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, Inject, Output, PLATFORM_ID } from '@angular/core';
import { FFResizeSensorService } from './ff-resize-sensor.service';
var FFResizeSensorDirective = /** @class */ (function () {
    function FFResizeSensorDirective(el, service, platformId) {
        this.el = el;
        this.service = service;
        this.platformId = platformId;
        this.resize = new EventEmitter();
    }
    /**
     * @return {?}
     */
    FFResizeSensorDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.service.subscribe(this.el.nativeElement, (/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            _this.onResize(event);
        }));
    };
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    FFResizeSensorDirective.prototype.onResize = /**
     * @private
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.oldWidth = event.oldWidth;
        this.oldHeight = event.oldHeight;
        this.width = event.width;
        this.height = event.height;
        this.resize.emit(event);
    };
    FFResizeSensorDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[ffResizeSensor]',
                    exportAs: 'ffResizeSensor'
                },] }
    ];
    /** @nocollapse */
    FFResizeSensorDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: FFResizeSensorService },
        { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
    ]; };
    FFResizeSensorDirective.propDecorators = {
        resize: [{ type: Output }]
    };
    return FFResizeSensorDirective;
}());
export { FFResizeSensorDirective };
if (false) {
    /** @type {?} */
    FFResizeSensorDirective.prototype.width;
    /** @type {?} */
    FFResizeSensorDirective.prototype.height;
    /** @type {?} */
    FFResizeSensorDirective.prototype.oldWidth;
    /** @type {?} */
    FFResizeSensorDirective.prototype.oldHeight;
    /** @type {?} */
    FFResizeSensorDirective.prototype.resize;
    /** @type {?} */
    FFResizeSensorDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    FFResizeSensorDirective.prototype.service;
    /**
     * @type {?}
     * @private
     */
    FFResizeSensorDirective.prototype.platformId;
}
