// ff-resize-sensor/esm2015/lib/ff-resize-sensor.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, Inject, Output, PLATFORM_ID } from '@angular/core';
import { FFResizeSensorService } from './ff-resize-sensor.service';
export class FFResizeSensorDirective {
    /**
     * @param {?} el
     * @param {?} service
     * @param {?} platformId
     */
    constructor(el, service, platformId) {
        this.el = el;
        this.service = service;
        this.platformId = platformId;
        this.resize = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.service.subscribe(this.el.nativeElement, (/**
         * @param {?} event
         * @return {?}
         */
        event => {
            this.onResize(event);
        }));
    }
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    onResize(event) {
        this.oldWidth = event.oldWidth;
        this.oldHeight = event.oldHeight;
        this.width = event.width;
        this.height = event.height;
        this.resize.emit(event);
    }
}
FFResizeSensorDirective.decorators = [
    { type: Directive, args: [{
                selector: '[ffResizeSensor]',
                exportAs: 'ffResizeSensor'
            },] }
];
/** @nocollapse */
FFResizeSensorDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: FFResizeSensorService },
    { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
];
FFResizeSensorDirective.propDecorators = {
    resize: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    FFResizeSensorDirective.prototype.width;
    /** @type {?} */
    FFResizeSensorDirective.prototype.height;
    /** @type {?} */
    FFResizeSensorDirective.prototype.oldWidth;
    /** @type {?} */
    FFResizeSensorDirective.prototype.oldHeight;
    /** @type {?} */
    FFResizeSensorDirective.prototype.resize;
    /** @type {?} */
    FFResizeSensorDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    FFResizeSensorDirective.prototype.service;
    /**
     * @type {?}
     * @private
     */
    FFResizeSensorDirective.prototype.platformId;
}
